### ISSUE TITLE
Simplify the counter example?

### ISSUE THREAD

**Issue by dmitriz** [funkia/turbine#50]
This seem to look like the case for the `loop`:
https://github.com/funkia/turbine/blob/master/examples/counters/src/index.ts#L44

That would remove the need for the model.

The component itself looks like a general purpose selector component,
to which you pass the array `labels = [1, 2, 3, 4]`,
along with the child component `selectorButton`.

I wonder if that can be made into something nice and reusable...

**Comment by dmitriz** [comment]
Here it might be nice to slightly overload the `div`, so you could simply pass the stream or behavior of component directly to `div`:

https://github.com/funkia/turbine/blob/master/examples/counters/src/index.ts#L59

I've thought that was also what happened here:
```js
...
const {inputValue: email} = yield input();
const isValid = email.map(isValidEmail);
yield div([
    "The address is ", map((b) => b ? "valid" : "invalid", isValid)
  ]);
```

The `isValid` behavior is directly passed to the `div`, without the `dynamic` call. Is it correct?

**Comment by dmitriz** [comment]
Also wouldn't say no to some more details on the "mysterious" `list` function :)

https://github.com/funkia/turbine/blob/master/examples/counters/src/version3.ts#L66

**Comment by paldepind** [comment]
@dmitriz 

> That would remove the need for the model, right?

Yes, it actually would. But in the next version of Hareactive `stepper` will return either a `Behavior<Behavior<A>>` or a `Now<Behavior<A>>` so then you'd need the model.

> The component itself looks like a general purpose selector component,
to which you pass the array `labels = [1, 2, 3, 4]`,
along with the child component `selectorButton`.
>
> I wonder if that can be made into something nice and reusable...

I think that is a good observation. It probably can be made into a completely reusable component.

> Here it might be nice to slightly overload the `div`, so you could simply pass the stream or behavior of component directly to `div`:
> 
> https://github.com/funkia/turbine/blob/master/examples/counters/src/index.ts#L59

Yes. It should already be possible to do that. I don't know why a put an explicit `dynamic` in the code.

> The `isValid` behavior is directly passed to the `div`, without the `dynamic` call. Is it correct?

Absolutely.

> Also wouldn't say no to some more details on the "mysterious" `list` function :)

I'll update #53 with documentation for both `lift` and `dynamic` :smile:

**Comment by dmitriz** [comment]
@paldepind 

>>That would remove the need for the model, right?

>Yes, it actually would. But in the next version of Hareactive stepper will return either a Behavior<Behavior<A>> or a Now<Behavior<A>> so then you'd need the model.

I would also like to avoid the `stepper` :)
I would add the memory behavior to the stream like in `flyd` as suggested here: https://github.com/funkia/hareactive/issues/29

It is really uniquely determined by the stream itself, so the structure is the same.
And has the benefit of being more convenient to anyone using `flyd` and 
to cut down the number of methods and reduce the code complexity, I would think.

Any reasons against it?

**Comment by dmitriz** [comment]
>>The component itself looks like a general purpose selector component,
to which you pass the array labels = [1, 2, 3, 4],
along with the child component selectorButton.
>>I wonder if that can be made into something nice and reusable...

> I think that is a good observation. It probably can be made into a completely reusable component.

I'd be really interested in that. 
There are very few such truly reusable components out there, as far I can see.
As @jayrbolton, it should be decoupled from the markup as much as possible.

On the most abstract level you select an element in array, so basically its index.
Every element in array should be a component outputting its own select stream.
These can be merged together and completed to the memory behavior that is fed back
to the selector.

On the outside, you pass it array of stateful components `compArray`,
each component parametrized by whether it is selected
(which can be just setting the `active` class but does not limit to it).
Basically, the whole array depends on the selection value, so is really a function
```js
compArray = (selectedIndex: Number): Array<Component>
```
that would be passed to the reusable `Selector` component as argument
```js
const { selected } = yield Selector(compArray)
```
with the `selected` stream/behavior output as in the example.

What do you think?

**Comment by dmitriz** [comment]
Another more restrictive alternative would be the array of selected/unselected component pairs:

```js
const arrayOfSelectableComponents: Array<(selected: Boolean): Component>
```

That is more focused to the common use case, 
where each component's state is its own selector.
